eclair-core/src/main/scala/fr/acinq/eclair/io/PendingChannelsRateLimiter.scala

```diff
@@ -0,0 +1,154 @@
+          case WrappedGetNodeResponse(_, UnknownNode(nodeId), _) =>
+            restoring(channels.tail, pendingPublicNodeChannels, pendingPrivateNodeChannels + (nodeId -> pendingChannels.map(_.channelId)))
+        }
+      case None =>
+        context.system.eventStream ! EventStream.Subscribe(context.messageAdapter[ChannelIdAssigned](c => ReplaceChannelId(c.remoteNodeId, c.temporaryChannelId, c.channelId)))
+        context.system.eventStream ! EventStream.Subscribe(context.messageAdapter[ChannelOpened](c => RemoveChannelId(c.remoteNodeId, c.channelId)))
+        context.system.eventStream ! EventStream.Subscribe(context.messageAdapter[ChannelClosed](c => RemoveChannelId(c.commitments.remoteNodeId, c.channelId)))
+        context.system.eventStream ! EventStream.Subscribe(context.messageAdapter[ChannelAborted](c => RemoveChannelId(c.remoteNodeId, c.channelId)))
+        registering(pendingPublicNodeChannels, pendingPrivateNodeChannels)
+    }
+  }
+
+  private def registering(pendingPublicNodeChannels: Map[PublicKey, Seq[ByteVector32]], pendingPrivateNodeChannels: Map[PublicKey, Seq[ByteVector32]]): Behavior[Command] = {
+    Metrics.OpenChannelRequestsPending.withTag(Tags.PublicPeers, value = true).update(pendingPublicNodeChannels.flatMap(_._2).size)
+    Metrics.OpenChannelRequestsPending.withTag(Tags.PublicPeers, value = false).update(pendingPrivateNodeChannels.size)
+    Behaviors.receiveMessagePartial {
+      case AddOrRejectChannel(replyTo, remoteNodeId, _) if nodeParams.channelConf.channelOpenerWhitelist.contains(remoteNodeId) =>
+        replyTo ! AcceptOpenChannel
+        Behaviors.same
+      case AddOrRejectChannel(replyTo, remoteNodeId, temporaryChannelId) =>
+        val adapter = context.messageAdapter[Router.GetNodeResponse](r => WrappedGetNodeResponse(temporaryChannelId, r, Some(replyTo)))
+        router ! GetNode(adapter, remoteNodeId)
+        Behaviors.same
+      case WrappedGetNodeResponse(temporaryChannelId, PublicNode(announcement, _, _), Some(replyTo)) =>
+        pendingPublicNodeChannels.get(announcement.nodeId) match {
+          case Some(pendingChannels) if pendingChannels.size >= nodeParams.channelConf.maxPendingChannelsPerPeer =>
+            replyTo ! ChannelRateLimited
+            Behaviors.same
+          case Some(peerChannels) =>
+            replyTo ! AcceptOpenChannel
+            registering(pendingPublicNodeChannels + (announcement.nodeId -> (temporaryChannelId +: peerChannels)), pendingPrivateNodeChannels)
+          case None =>
+            replyTo ! AcceptOpenChannel
+            registering(pendingPublicNodeChannels + (announcement.nodeId -> Seq(temporaryChannelId)), pendingPrivateNodeChannels)
+        }
+      case WrappedGetNodeResponse(temporaryChannelId, UnknownNode(nodeId), Some(replyTo)) =>
+        if (pendingPrivateNodeChannels.flatMap(_._2).size >= nodeParams.channelConf.maxTotalPendingChannelsPrivateNodes) {
```

Comment:
nit: this creates a new list from each private node's channels list just to get the size, which is quite wasteful. The following code should be better:

```suggestion
if (pendingPrivateNodeChannels.map(_._2.length).sum >= nodeParams.channelConf.maxTotalPendingChannelsPrivateNodes) {
```

Reply:
Good idea; it's fixed in [9b27eed](https://github.com/ACINQ/eclair/pull/2552/commits/9b27eed4003dde3e8440f43a3d87bd548f95b058)